src/state/delete-text.action.ts

```diff
@@ -0,0 +1,185 @@
+import { Action } from './action';
+import { ISegment } from '../entities/segment';
+import { ICursor } from '../entities/cursor';
+import { getCurrSegment, IEditorState, findSegment } from './editor.state';
+import { remove } from '../common/array';
+
+export class DeleteTextAction implements Action {
+
+    private rematchIndex(segments: ISegment[]): ISegment[]{
+        return segments.map((currSegment: ISegment, idx: number) => {
+            return Object.assign<any, ISegment, Partial<ISegment>>(
+                {},
+                currSegment,
+                {
+                    index: idx
+                }
+            );
+        });
+    }
+
+    private deleteSingleChar(state: IEditorState): IEditorState {
+        const { cursor, segments } = state;
+        let newState: IEditorState = state;
+        let newSegments: ISegment[] = segments;
+        let newCursor: ICursor = cursor;
+
+        const segment = getCurrSegment(newState);
+        if (!segment) {
+            return state;
+        }
+
+        const newSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            segment,
+            {
+                // to remove the single char at the cursor
+                content: remove<string>(
+                    segment.content,
+                    newCursor.startOffset - 1
+                )
+            }
+        );
+
+        newSegments = newSegments.map((currSegment: ISegment) => {
+            if (currSegment.index === segment.index) {
+                return newSegment;
+            }
+            return currSegment;
+        });
+
+        // If after the deletion, the segment is empty, we need to remove it, and set the cursor to the end of the previous segment
+        if (newSegment.content.length === 0) {
+            newSegments = remove<ISegment>(newSegments, newSegment.index);
+
+            // set the cursor to the end of the previous segment
+            if (newSegment.index > 0) {
+                let prevSegment = findSegment(segment.index - 1, segments);
+
+                newCursor = Object.assign({}, newCursor, {
+                    startSegmentIndex: prevSegment.index,
+                    startOffset: prevSegment.content.length,
+                    endSegmentIndex: prevSegment.index,
+                    endOffset: prevSegment.content.length
+                });
+
+                // rematch index here since we have removed a segment
+                newSegments = this.rematchIndex(newSegments);
+
+            // If it's the first segment, we need to reset the cursor.
+            } else {
+                newCursor = Object.assign({}, newCursor, {
+                    startSegmentIndex: 0,
+                    startOffset: 0,
+                    endSegmentIndex: 0,
+                    endOffset: 0
+                });
+            }
```

Comment:
```suggestion
           {newSegments, newCursor} = this.delectEmptySegment(newSegment);
```

Reply:
Let's move this inside `delectEmptySegment`. This will reduce the number of nest `if else` statement in this function to simplify the logic. Inside `delectEmptySegment`, we then only need one `if` statement to deal with `cursor` jumping.